[build] Run Venice CI with GitHub Actions

sushantmane opened the pull request:
Run Venice tests on the following matrix:
OS: macOS, Ubuntu
JDK: 11

<!--
Add a list of affected components in the PR title in the following format:
[component1]...[componentN] Concise commit message

Valid component tags are: [da-vinci], [server], [controller],
[router], [samza], [vpj], [fast-client], [thin-client], [alpini],
[admin-tool], [test], [build], [doc], [script]

Example title: [server][da-vinci] Use dedicated thread to persist data to storage engine

Note: PRs with titles not following the format will not be merged
-->

## Run Venice CI with GitHub Actions
<!--
Describe
- What changes to make and why you are making these changes.
- How are you going to achieve your goal.
- Describe what testings you have done, for example, performance testing etc.

If this PR resolves an issue be sure to include "Resolves #XXX" to correctly link and close the issue upon merge.
-->
While we figure out how we want to run CI with GitHub actions, I think it would be good to have
something up and running in the meantime. With this RB, CI will be run on every commit/pull request. 
It triggers one job/OS for unit tests (~42 mins) and one for integration tests/OS (2.30 hours and a bit flaky). 




## How was this PR tested?
<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask. We'd be happy to help.
-->

## Does this PR introduce any user-facing changes?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, choose 'No'.
-->
- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

FelixGV commented:
Thanks Sushant! Is it normal that the checks show as failed?

sushantmane commented:
> Thanks Sushant! Is it normal that the checks show as failed?

No, checks should not fail. I know the reason: this PR ('get fetch' part in actions) was designed to work on forks and I need to update it to work everywhere. I'll update it. Thanks.

nisargthakkar commented on clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/utils/IsolatedIngestionUtils.java:
```diff
@@ -308,7 +308,11 @@ public static void releaseTargetPortBinding(int port) {
    */
   public static Optional<Integer> getLingeringIngestionProcessId(int port) {
     Optional<Integer> isolatedIngestionProcessPid = Optional.empty();
-    String processIds = executeShellCommand("/usr/sbin/lsof -t -i :" + port);
```
Why do we need the `/usr/sbin` prefix? It should be in the path by default

sushantmane replied:
A full path is required for CentOS/RHEL. The path is not the same on Ubuntu and it works without the path on it.  

Also, GH actions runner doesn't have CentOS and hence I'm using Ubuntu.

nisargthakkar commented on .github/workflows/gh-ci.yml:
```diff
@@ -0,0 +1,88 @@
+        with:
+          arguments: "clean check -x spotlessCheck --continue --no-daemon -Pspotallbugs -Pspotbugs.xml -Pspotbugs.ignoreFailures -x integrationTest -x test"
+
+  UnitTests:
```
Can you also add a spotbugs step?

nisargthakkar replied:
I see it's already there. Removing

FelixGV commented:
Hi Sushant, would you mind rebasing and retrying this? Curious to see how it goes now. Thanks!

FelixGV commented:
I think we need to divide the e2e tests further. One first cut might be to do a split along the lines of:

1. Run all integration tests that match `com.linkedin.venice.endToEnd.*`.
2. Run all integration tests that don't match `com.linkedin.venice.endToEnd.*`

Might need to define a couple more test targets in the Gradle build. This seems useful for that: https://mkyong.com/gradle/gradle-how-to-exclude-some-tests/

Are you game to give it a try?

sushantmane commented:
> I think we need to divide the e2e tests further. One first cut might be to do a split along the lines of:
> 
> 1. Run all integration tests that match `com.linkedin.venice.endToEnd.*`.
> 2. Run all integration tests that don't match `com.linkedin.venice.endToEnd.*`
> 
> Might need to define a couple more test targets in the Gradle build. This seems useful for that: https://mkyong.com/gradle/gradle-how-to-exclude-some-tests/
> 
> Are you game to give it a try?

> 

Sorry for the late response. Sure, I'll split e2e tests into two parts. They used to finish in [< 3HR ](https://github.com/sushantmane/venice/actions/runs/3521681573/jobs/5903795730) and hence I added 3HR timeout. But it seems we've added >25 e2e tests since then.

sushantmane commented:
@FelixGV - 
> 1. Run all integration tests that match com.linkedin.venice.endToEnd.*.

Completes in 2h10m

> 2. Run all integration tests that don't match com.linkedin.venice.endToEnd.*

Completes in 1h

@adamxchen  is planning to work on this in the next quarter. In the meantime, should we use this PR?

FelixGV commented:
Thanks a lot for the initiative Sushant!! I think we should merge this in, for sure!

In the future, I would like us to reduce the wall clock time by splitting further, and/or deleting redundant tests.

In any case, this is a great start! Thanks again 🙏

FelixGV reviewed (approved):
Tremendous 🤩🍾🙏

sushantmane commented:
Thanks for the review. I'll run our internal CI once just to make sure everything else is alright.

sushantmane commented:
@FelixGV - Could you please approve the PR again?

nisargthakkar reviewed (approved):
LGTM

FelixGV reviewed (approved):
Thanks a lot! Please make sure to review and clean up the commit message while squashing.

sushantmane commented:
Thank you @FelixGV & @nisargthakkar!
